python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -25,6 +26,118 @@
 
 
+class NestedTorchTensorNcclChannel(ChannelInterface):
+    def __init__(
+        self,
+        writer: ray.actor.ActorHandle,
+        readers: List[ray.actor.ActorHandle],
+        gpu_data_typ: "TorchTensorType",
+        cpu_data_typ: Optional["SharedMemoryType"] = None,
+        _gpu_data_channel: Optional["TorchTensorNcclChannel"] = None,
+        _cpu_data_channel: Optional["Channel"] = None,
+    ):
+        self._writer = writer
+        self._readers = readers
+
+        if _gpu_data_channel is not None or _cpu_data_channel is not None:
+            self._gpu_data_channel = _gpu_data_channel
+            self._cpu_data_channel = _cpu_data_channel
+        else:
+            self._gpu_data_channel: TorchTensorNcclChannel = (
+                gpu_data_typ.create_channel(writer, readers)
+            )
+            self._cpu_data_channel: Optional["Channel"] = None
+            if cpu_data_typ is not None:
+                self._cpu_data_channel = cpu_data_typ.create_channel(writer, readers)
+
+        # Used for serialization.
+        self._worker = ray._private.worker.global_worker
+        self._worker.check_connected()
+
+        ctx = ChannelContext.get_current()
+        self.serialization_ctx = ctx.serialization_context
+        assert self.serialization_ctx is not None
+
+    @classmethod
+    def from_channels(
+        cls,
+        gpu_data_channel: "TorchTensorNcclChannel",
+        cpu_data_channel: Optional["Channel"],
+    ):
+        return cls(
+            writer=None,
+            readers=None,
+            gpu_data_typ=None,
+            cpu_data_typ=None,
+            _gpu_data_channel=gpu_data_channel,
+            _cpu_data_channel=cpu_data_channel,
+        )
+
+    def __reduce__(self):
+        return (
+            NestedTorchTensorNcclChannel.from_channels,
+            (self._gpu_data_channel, self._cpu_data_channel),
+        )
+
+    def ensure_registered_as_writer(self):
+        self._gpu_data_channel.ensure_registered_as_writer()
+        if self._cpu_data_channel is not None:
+            self._cpu_data_channel.ensure_registered_as_writer()
+
+    def ensure_registered_as_reader(self):
+        self._gpu_data_channel.ensure_registered_as_reader()
+        if self._cpu_data_channel is not None:
+            self._cpu_data_channel.ensure_registered_as_reader()
+
+    def write(self, value: Any):
+        self.serialization_ctx.reset_tensors([])
+
+        try:
+            serialized_cpu_data = self._worker.get_serialization_context().serialize(
+                value
+            )
+        except TypeError as e:
+            sio = io.StringIO()
+            ray.util.inspect_serializability(value, print_file=sio)
+            msg = (
+                "Could not serialize the put value "
+                f"{repr(value)}:\n"
+                f"{sio.getvalue()}"
+            )
+            raise TypeError(msg) from e
+
+        tensors_to_send = self.serialization_ctx.reset_tensors([])
```

Comment:
Not related to this PR:

The ChannelOutputType calls `ray.util.serialization.register_serializer` to register the serializer / deserializer. When / Where are the serializer / deserializer used?

```python
ray.util.serialization.register_serializer(
    torch.Tensor,
    serializer=serialize,
    deserializer=deserialize,
)
```

The tensors appear to be serialized and stored in `serialization_ctx.tensors` before this write function is called. We serialize CPU data in this function, but serialize tensors in other places. Why do we choose to serialize them in different locations?

Reply:
`register_serializer` registers a custom serializer for any objects that match the given type. So when we call the ray serializer with `self._worker.get_serialization_context().serialize(value)`, we will trigger the custom `deserialize` function for any torch.Tensors found in `value`. That will replace the tensors with the placeholders and add them to `serialization_ctx.tensors`.